hadoop-tools/hadoop-aws/src/test/java/org/apache/hadoop/fs/s3a/impl/ITestS3APutIfMatch.java

```diff
@@ -0,0 +1,150 @@
+
+    @Override
+    public Configuration createConfiguration() {
+        Configuration conf = super.createConfiguration();
+        S3ATestUtils.disableFilesystemCaching(conf);
+        removeBaseAndBucketOverrides(
+                conf,
+                MULTIPART_SIZE,
+                UPLOAD_PART_COUNT_LIMIT,
+                MIN_MULTIPART_THRESHOLD);
+        conf.setLong(UPLOAD_PART_COUNT_LIMIT, 2);
+        conf.setLong(MIN_MULTIPART_THRESHOLD, MULTIPART_MIN_SIZE);
+        conf.setInt(MULTIPART_SIZE, MULTIPART_MIN_SIZE);
+        return conf;
+    }
+
+    @Override
+    public void setup() throws Exception {
+        super.setup();
+        conf = createConfiguration();
+        skipIfNotEnabled(conf, FS_S3A_CONDITIONAL_FILE_CREATE,
+                "Skipping IfNoneMatch tests");
+    }
+
+    private static void assertS3ExceptionStatusCode(int code, Exception ex) {
+        S3Exception s3Exception = (S3Exception) ex.getCause();
+
+        if (s3Exception.statusCode() != code) {
+            throw new AssertionError("Expected status code " + code + " from " + ex, ex);
+        }
+    }
+
+    protected String getBlockOutputBufferName() {
+        return FAST_UPLOAD_BUFFER_ARRAY;
+    }
+
+    /**
+     * Create a file using the PutIfMatch feature from S3
+     * @param fs filesystem
+     * @param path       path to write
+     * @param data source dataset. Can be null
+     * @throws IOException on any problem
+     */
+    private static void createFileWithIfNoneMatchFlag(
+            FileSystem fs,
+            Path path,
+            byte[] data,
+            String ifMatchTag) throws Exception {
+        FSDataOutputStreamBuilder builder = fs.createFile(path);
+        builder.must(FS_S3A_CONDITIONAL_FILE_CREATE, "true");
+        builder.opt(FS_S3A_CREATE_HEADER + "." + IF_NONE_MATCH, ifMatchTag);
+        FSDataOutputStream stream = builder.create().build();
+        if (data != null && data.length > 0) {
+            stream.write(data);
+        }
+        stream.close();
+        IOUtils.closeStream(stream);
+    }
+
+    @Test
+    public void testPutIfAbsentConflict() throws Throwable {
+        FileSystem fs = getFileSystem();
+        Path testFile = methodPath();
+        fs.mkdirs(testFile.getParent());
+        byte[] fileBytes = dataset(TEST_FILE_LEN, 0, 255);
+
+        createFileWithIfNoneMatchFlag(fs, testFile, fileBytes, "*");
+
+        RemoteFileChangedException firstException = intercept(RemoteFileChangedException.class,
+                () -> createFileWithIfNoneMatchFlag(fs, testFile, fileBytes, "*"));
+        assertS3ExceptionStatusCode(412, firstException);
+
+        RemoteFileChangedException secondException = intercept(RemoteFileChangedException.class,
+                () -> createFileWithIfNoneMatchFlag(fs, testFile, fileBytes, "*"));
+        assertS3ExceptionStatusCode(412, secondException);
+    }
+
+
+    @Test
+    public void testPutIfAbsentLargeFileConflict() throws Throwable {
```

Comment:
I'd offered a way to test this without needing to create a large file for a multipart upload...we can work on this...anything uploading 6 MB gets moved into the scale tests, so doesn't get run as often